Harfbuzz fails to build on Ubuntu 20.04 (WSL 2)

# Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.

- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

<!-- To help us debug your issue, please complete these sections: -->

## What you were trying to do (and why)

I tried updating my exisiting `harfbuzz` installation.

## What happened (include command output)

After `harfbuzz` was patched, building with `ninja` failed

<details>
  <summary>Command output</summary>
  <pre>

  $ brew upgrade

==> Upgrading 1 outdated package:
harfbuzz 2.6.8 -> 2.7.0
==> Upgrading harfbuzz 2.6.8 -> 2.7.0
==> Downloading https://github.com/harfbuzz/harfbuzz/commit/7c61caa7384e9c3afa0d9237bf6cd303eb5ef3a1.patch?full_index=1
Already downloaded: /home/vsong/.cache/Homebrew/downloads/c06e429357a688c79e263d2094e103a4a9b969196ed5bcb6de9e13b6cc9698fd--7c61caa7384e9c3afa0d9237bf6cd303eb5ef3a1.patch
==> Downloading https://github.com/harfbuzz/harfbuzz/raw/fc0daafab0336b847ac14682e581a8838f36a0bf/test/shaping/fonts/sha1sum/270b89df543a7e48e206a2d830c0e10e5265c630.ttf
Already downloaded: /home/vsong/.cache/Homebrew/downloads/b0f20064588dc70dd5c74e36b842326d9dbbeacab35c63c3d900a8b34aaba98f--270b89df543a7e48e206a2d830c0e10e5265c630.ttf
==> Downloading https://github.com/harfbuzz/harfbuzz/archive/2.7.0.tar.gz
Already downloaded: /home/vsong/.cache/Homebrew/downloads/375e0a0893ea76812873fd49ef3bb2d3f178a39933235ffb8393748fbf388f69--harfbuzz-2.7.0.tar.gz
==> Patching
==> Applying 7c61caa7384e9c3afa0d9237bf6cd303eb5ef3a1.patch
patching file src/meson.build
Hunk #1 succeeded at 397 with fuzz 1 (offset -8 lines).
Hunk #2 succeeded at 406 (offset -8 lines).
Hunk #3 succeeded at 430 (offset -8 lines).
Hunk #4 succeeded at 532 (offset -8 lines).
Hunk #5 succeeded at 618 (offset -8 lines).
==> meson --prefix=/home/linuxbrew/.linuxbrew/Cellar/harfbuzz/2.7.0 --libdir=/home/linuxbrew/.linuxbrew/Cellar/harfbuzz/2.7.0/lib --buildtype=release --default-library=both -Dcairo=enabled -Dcoretext=enabled -Dfreetype=ena
==> ninja
Last 15 lines from /home/vsong/.cache/Homebrew/Logs/harfbuzz/02.ninja:
[115/289] Compiling C++ object test/api/test-cplusplus.p/test-cplusplus.cc.o
[116/289] Compiling C object test/api/test-ot-ligature-carets.p/test-ot-ligature-carets.c.o
[117/289] Compiling C object test/api/test-common.p/test-common.c.o
[118/289] Compiling C object test/api/test-map.p/test-map.c.o
[119/289] Generating hb-gobject-enums-tmp.cc with a meson_exe.py custom command
[120/289] Compiling C object test/api/test-buffer.p/test-buffer.c.o
[121/289] Compiling C++ object util/hb-ot-shape-closure.p/options.cc.o
[122/289] Compiling C object test/api/test-font.p/test-font.c.o
[123/289] Compiling C object test/api/test-object.p/test-object.c.o
[124/289] Compiling C++ object src/test-iter.p/test-iter.cc.o
[125/289] Compiling C++ object src/test-bimap.p/test-bimap.cc.o
[126/289] Compiling C++ object src/libharfbuzz.so.0.20700.0.p/hb-ot-layout.cc.o
[127/289] Compiling C++ object src/libharfbuzz-subset.so.0.20700.0.p/hb-subset-plan.cc.o
[128/289] Compiling C++ object src/libharfbuzz-subset.so.0.20700.0.p/hb-subset.cc.o
ninja: build stopped: subcommand failed.

READ THIS: https://docs.brew.sh/Troubleshooting
  </pre>
</details>
    
<details>
<summary>brew config</summary>

<pre>
HOMEBREW_VERSION: 2.4.9-218-g7075959
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7075959919691b9debac3d1bf00b5b257eb036a9
Last commit: 3 hours ago
Core tap ORIGIN: https://github.com/Homebrew/linuxbrew-core
Core tap HEAD: ff8d5de9caecfa4f75406990343cb222ec8e67a5
Core tap last commit: 9 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 16
CPU: 16-core 64-bit skylake
Homebrew Ruby: 2.6.3 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
Clang: N/A
Git: 2.25.1 => /bin/git
Curl: 7.68.0 => /usr/bin/curl
Kernel: Linux 4.19.104-microsoft-standard x86_64 GNU/Linux
OS: Ubuntu 20.04.1 LTS (focal)
Host glibc: 2.31
/usr/bin/gcc: 9.3.0
glibc: N/A
gcc: 5.5.0_7
xorg: 20170115_1
</pre>
</details>

<details>
<summary>brew doctor</summary>
<pre>
Your system is ready to brew.
</pre>
</details>

<details>
<summary>brew gist-logs harfbuzz</summary>
<pre>
https://gist.github.com/Tyrubias/e1d506f78672f29cfeb6ff6029839d39
</pre>
</details>

## What you expected to happen

Building of `harfbuzz` should have succeeded.

## Step-by-step reproduction instructions (by running `brew install` commands)

`brew upgrade harfbuzz`

```
glib-2.0/glib/gatomic.h:128:15: error: variable 'gapg_temp_atomic' set but not used [-Werror=unused-but-set-variable]
     gpointer *gapg_temp_atomic = (gpointer *)(atomic);                       \
               ^
```

This is an upstream bug: https://github.com/harfbuzz/harfbuzz/issues/2555

--- a/Formula/harfbuzz.rb
+++ b/Formula/harfbuzz.rb
@@ -35,6 +35,10 @@ class Harfbuzz < Formula
     sha256 "4d60b681918c2b911da9a84e37386ed1fa48794d5c943ff9a7bd50eb3a255969"
   end
 
+  # Silence a warning that prevents building on Linux
+  # https://github.com/harfbuzz/harfbuzz/issues/2555
+  patch :DATA
+
   def install
     args = %w[
       --default-library=both
@@ -62,3 +66,29 @@ class Harfbuzz < Formula
     end
   end
 end
+
+__END__
+diff --git a/src/hb-gobject-enums.cc.tmpl b/src/hb-gobject-enums.cc.tmpl
+index 2ffd1c9d..26954ed2 100644
+--- a/src/hb-gobject-enums.cc.tmpl
++++ b/src/hb-gobject-enums.cc.tmpl
+@@ -30,6 +30,7 @@
+ #ifdef HAVE_GOBJECT
+
+ /* g++ didn't like older gtype.h gcc-only code path. */
++#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
+ #include <glib.h>
+ #if !GLIB_CHECK_VERSION(2,29,16)
+ #undef __GNUC__
+diff --git a/src/hb-gobject-structs.cc b/src/hb-gobject-structs.cc
+index 7c46e264..a64913a9 100644
+--- a/src/hb-gobject-structs.cc
++++ b/src/hb-gobject-structs.cc
+@@ -50,6 +50,7 @@
+
+
+ /* g++ didn't like older gtype.h gcc-only code path. */
++#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
+ #include <glib.h>
+ #if !GLIB_CHECK_VERSION(2,29,16)
+ #undef __GNUC__

